CI: Select jobs by touched code

ldoktor opened the pull request:
This PR adds 2 workflows:

* gatekeeper-skipper
* gatekeeper

The skipper can be used in workflows to detect changes and output set of tests to be executed (or skipped), the gatekeeper watches all the running jobs and checks the status of expected jobs to pass (based on changed files), which should replace the "required" feature of GH (once we make this workflow required).

Note this solution should be quite scalable so things like advanced heuristic or GH comments/labels overrides should be possible.

Example PRs (with slightly outdated gatekeeper but mainly similar):

* https://github.com/ldoktor/kata-containers/pull/6
* https://github.com/ldoktor/kata-containers/pull/5

Note you can get gatekeeper output of any existing PR (yours or kata-containers) by running something like `GITHUB_TOKEN="" REQUIRED_JOBS="foo;bar" REQUIRED_REGEXPS=".*" COMMIT_HASH=b8382cea886ad9a8f77d237bcfc0eba0c98775dd GITHUB_REPOSITORY=kata-containers/kata-containers python3 tools/testing/gatekeeper/jobs.py`

wainersm commented on tools/testing/gatekeeper/skips.py:
```diff
@@ -0,0 +1,102 @@
+                if regexp.search(line):
+                    for feature in features:
+                        enabled_features.add(feature)
+                    break
```
Hi @ldoktor !

This `break` is very important, as it applies the `sort by order of importance` rule. For example,  if the first match suggest an empty list of features then, in practice, it is a skip; if there are other matches to this same line that suggest a list of features then they won't apply. So shall we get this line document to avoid future bugs by accident removal of this `break` or even a change on the algorithm?

ldoktor replied:
The same here, I'll add a comment in v2 (my code/comment ratio is way below my average as this is a product of many failed attempts)

wainersm commented on tools/testing/gatekeeper/skips.py:
```diff
@@ -0,0 +1,102 @@
+        enabled_features = self.get_features(target_branch)
+        if not tests:
+            for feature in self.all_set_of_tests:
+                print(f"skip_{feature}=" +
```
The output of this script is redirected (appended)  to `$GITHUB_OUTPUT` which end up as outputs of the skipper job. It's is expected the format `key=value` as in https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs . Maybe worth a inline documentation to, again, avoid changing this by accident.

ldoktor replied:
Sure, let me add it to v2

wainersm commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,189 @@
+            headers=_GH_HEADERS,
+            timeout=60
+        )
+        response.raise_for_status()
```
In two places it can crash this script (` response.raise_for_status()` throwing exception due API unavailable), failing the gatekeeper job and consequently blocking a PR. Can we resume the checking by re-running the gatekeeper job from the Github web UI?

Regardless, I think we should implement a retry mechanism to call the Github API. The interwebs says `requests` has a retry implementation with backoff!

ldoktor replied:
At this point I'd like to keep it this way to experiment with the limits. If it becomes an issue, we should get the exception with details. I haven't found the limits for jobs on GH runners but somewhere I read there should not be any limit.

Anyway it's just another workflow, one can re-try it from the UI.

wainersm replied:
That's a good idea. Let's see how it works in production then we fine tune.

wainersm commented on .github/workflows/gatekeeper.yaml:
```diff
@@ -0,0 +1,37 @@
+          fetch-depth: 0
+      - id: gatekeeper
+        env:
+          PR_NUMBER: ${{ github.event.pull_request.number }}
```
` PR_NUMBER` seems unused.

wainersm commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,189 @@
+                                    for job, status in self.results
+                                    if status == RUNNING])
+                print(f"{running_jobs} jobs are still running...")
+                time.sleep(60)
```
It would be good to parameterize this value, 60s seems too small (build and test jobs take several minutes to finish). I.e. we might need to tune that value to ensure we don't hit the API rate limits:

 * https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-github_token-in-github-actions
 * https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits

ldoktor replied:
Wow, you found the limits, good. We can change it as well as we can make it dynamic (let's say we expect the failures first after 15 or 30m, then we can start checking per 1m and after 2h increase it to hour (the assumption would be we know tests don't finish earlier than in 30m, then we want to be checking frequently to fail fast but when things take too long, we don't need that frequency and can check once an hour). Or anything else we can come-up with.

ldoktor replied:
Anyway, I bumped it to 3m for now, let me know if you want some advanced heuristic

wainersm replied:
Initially 3m seems fine. let's see how it behave in practice ;)

wainersm commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,189 @@
+                print(f"{running_jobs} jobs are still running...")
+                time.sleep(60)
+                continue
+            sys.exit(ret)
```
One improvement that I'm still unsure if it should implemented now or later, is about proper reporting the required jobs that failed. With this new approach we won't have the '(Required)' marker in front of the jobs and I'm afraid users (mainly newcomers) won't immediately open the gatekeeper's logs to check the jobs that failed.

And we might be on the edge for creating a Github App :)

ldoktor replied:
They would get a bunch of failed non-required tests but one (or multiple) failed required tests, one of them being the gatekeeper. And last note from gatekeeper should explain the situation and eventually navigate them to the right place.

wainersm replied:
ok

pmores commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,189 @@
+* COMMIT_HASH: Full commit hash we want to be watching
+* GITHUB_REPOSITORY: Github repository (user/repo)
+Sample execution (GH token can be excluded):
+GITHUB_TOKEN="..." REQUIRED_JOBS="skipper / skipper"
```
Apologies in advance if I'm just being boneheaded but line 10 above says that `REQUIRED_JOBS` should be a comma-separated list which `"skipper / skipper"` doesn't seem to be.

ldoktor replied:
It is, the full name of skipper job is `skipper / skipper` as it includes the `workflow_name`.

zvonkok commented:
Would this also help in rerunning only failed jobs rather than rerunning a complete workflow?

ldoktor commented:
Changes:

* removed unused "PR_NUMBER" env variable
* better description for REQUIRED_JOBS ("workflow / job")
* decreased checking period (1m -> 3m sleep)
* 2 extra comments in `skips.py` to emphasize the right format and order

stevenhorsman commented:
> Note you can get gatekeeper output of any existing PR (yours or kata-containers) by running something like `GITHUB_TOKEN="" REQUIRED_JOBS="foo;bar" REQUIRED_REGEXPS=".*" COMMIT_HASH=b8382cea886ad9a8f77d237bcfc0eba0c98775dd GITHUB_REPOSITORY=kata-containers/kata-containers python3 tools/testing/gatekeeper/jobs.py`

Hi @ldoktor - as part of reviewing this I tried to run the check on https://github.com/kata-containers/kata-containers/pull/9702 by running
```
REQUIRED_JOBS="Commit Message Check" REQUIRED_REGEXPS=".*" GITHUB_REPOSITORY=ildikov/kata-containers COMMIT_HASH=5e03bec26b5d40b48c60ceaa5eda5f533ea34842 python3 tools/testing/gatekeeper/jobs.py
 ```
It is probably user error, but I got a failure:
```
WARN: Commit Message Check - Still running

Some jobs are still running.
Traceback (most recent call last):
  File "/root/kata-containers/tools/testing/gatekeeper/jobs.py", line 190, in <module>
    Checker().run()
  File "/root/kata-containers/tools/testing/gatekeeper/jobs.py", line 181, in run
    for job, status in self.results
        ^^^^^^^^^^^
ValueError: too many values to unpack (expected 2)
```
Any idea what I did wrong?
Thanks!

stevenhorsman commented:
> Note you can get gatekeeper output of any existing PR (yours or kata-containers) by running something like GITHUB_TOKEN="" REQUIRED_JOBS="foo;bar" REQUIRED_REGEXPS=".*" COMMIT_HASH=b8382cea886ad9a8f77d237bcfc0eba0c98775dd GITHUB_REPOSITORY=kata-containers/kata-containers python3 tools/testing/gatekeeper/jobs.py

I also noticed the jobs reporting might need some work - when I copied your example it run successfully, but I had:
`WARN: foo;bar - Still running` in the output.
I'm not 100% sure if we should WARN, or FAIL when a required test isn't found, but is there a way we can distinguish between still running and doesn't exist?

stevenhorsman reviewed (commented):
I've had a first pass at reviewing and hit a couple of issues (mentioned in the PR) when manually testing it and a couple of code questions/comments

ldoktor commented on tools/testing/gatekeeper/required-tests.yaml:
```diff
@@ -0,0 +1,37 @@
+  # Mapping of path (python) regexps to set-of-tests (sort by order of importance)
+  # CI
+  - "^ci/openshift-ci/": []
+  - "^tools/testing/gatekeeper": []
```
This is a remnant from when I assumed we won't be using from-PR version of gatekeeper. Let me remove it.

ldoktor commented on .github/workflows/static-checks.yaml:
```diff
@@ -12,7 +12,16 @@ concurrency:
 name: Static checks
 jobs:
+  skipper:
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'ok-to-test') }}
```
True, I missed that, let me fix that in v2

ldoktor commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,190 @@
+        jobs = []
+        page = 1
+        while True:
+            response = requests.get(
```
I didn't wanted to rely on extra tools (not sure if they are available on workers). I can experiment with that.

stevenhorsman replied:
Yeah - I check first and `gh` is available on the runners (at least the 22.04 one). Sorry I should have included the link: https://github.com/actions/runner-images/blob/ubuntu22/20240304.1/images/ubuntu/Ubuntu2204-Readme.md#cli-tools

stevenhorsman replied:
I'm happy for this to be deferred to a later issue

ldoktor commented:
Changes:

* static-checks not to require ok-to-test label
* removed gatekeeper `required-tests` regexp as we might want to test all changes there
* rebased (mainly straight forward, only the qemu-runtime-rs was added to aks)

Not changed:

* The use of `gh`, let's experiment with pure python first, we can rework it with gh later

sprt commented on tools/testing/gatekeeper/required-tests.yaml:
```diff
@@ -0,0 +1,36 @@
+  #- "^src/": ["static", "build", "test"]
+
+mapping:
+  # Mapping of set-of-tests to required test names and/or test name regexps
```
I would simplify this and make each of `test/build/static` only take a list of regexes, since a name can easily be converted into a regex by escaping it.

ldoktor replied:
Well that depends on the usage. If we're likely to be using regexps only, then yes, but if there were many individual jobs people wanted to enumerate, then having it as a list should be easier to read and actually safer as it's clearly stated what "items" are expected. (underneath the regexps and names are handled completely differently, individual names can be expected in 1:1 mapping while regexps resolve into unknown amount of items).

With that said I'd prefer keeping it this way for now and once we have some real-world config we could refine, what do you think?

wainersm replied:
sounds reasonable to me.

wainersm commented:
Hi @zvonkok !

> Would this also help in rerunning only failed jobs rather than rerunning a complete workflow?

You asked this question on AC and @ldoktor explained already. To add my understanding is that as is today, it won't re-running failed tests, as devs still need to perform that operation on web UI. However, it opens the door for allowing us to implement to trigger actions based on commands, so that we could (in the future) implement a "/rerun-failed-tests".

wainersm commented:
> for job, status in self.results



> > Note you can get gatekeeper output of any existing PR (yours or kata-containers) by running something like `GITHUB_TOKEN="" REQUIRED_JOBS="foo;bar" REQUIRED_REGEXPS=".*" COMMIT_HASH=b8382cea886ad9a8f77d237bcfc0eba0c98775dd GITHUB_REPOSITORY=kata-containers/kata-containers python3 tools/testing/gatekeeper/jobs.py`
> 
> Hi @ldoktor - as part of reviewing this I tried to run the check on #9702 by running
> 
> ```
> REQUIRED_JOBS="Commit Message Check" REQUIRED_REGEXPS=".*" GITHUB_REPOSITORY=ildikov/kata-containers COMMIT_HASH=5e03bec26b5d40b48c60ceaa5eda5f533ea34842 python3 tools/testing/gatekeeper/jobs.py
> ```
> 
> It is probably user error, but I got a failure:
> 
> ```
> WARN: Commit Message Check - Still running
> 
> Some jobs are still running.
> Traceback (most recent call last):
>   File "/root/kata-containers/tools/testing/gatekeeper/jobs.py", line 190, in <module>
>     Checker().run()
>   File "/root/kata-containers/tools/testing/gatekeeper/jobs.py", line 181, in run
>     for job, status in self.results
>         ^^^^^^^^^^^
> ValueError: too many values to unpack (expected 2)
> ```
> 
> Any idea what I did wrong? Thanks!

Hi @stevenhorsman !

Wrong syntax. To read the results mapping it should `for job, status in self.results.items()`. I'm gonna fix it.

wainersm commented:
Pushed two fixes to @ldoktor 's tree:
 * rebased to main then fix conflict in .github/workflows/ci.yaml
   * To keep the garm workflows disabled, added a `false` to https://github.com/kata-containers/kata-containers/pull/9637/commits/d364b0fc136674288f1be4d14b2f7b48bfe500bb#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR121 and https://github.com/kata-containers/kata-containers/pull/9637/commits/d364b0fc136674288f1be4d14b2f7b48bfe500bb#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR135

wainersm commented:
> > Note you can get gatekeeper output of any existing PR (yours or kata-containers) by running something like GITHUB_TOKEN="" REQUIRED_JOBS="foo;bar" REQUIRED_REGEXPS=".*" COMMIT_HASH=b8382cea886ad9a8f77d237bcfc0eba0c98775dd GITHUB_REPOSITORY=kata-containers/kata-containers python3 tools/testing/gatekeeper/jobs.py
> 
> I also noticed the jobs reporting might need some work - when I copied your example it run successfully, but I had: `WARN: foo;bar - Still running` in the output. I'm not 100% sure if we should WARN, or FAIL when a required test isn't found, but is there a way we can distinguish between still running and doesn't exist?

Good question. @ldoktor will know for sure; but I believe it depends. With current implementation if you export `REQUIRED_JOBS` and the jobs doesn't exist then it will keep waiting the job conclusion forever (or hit a timeout somewhere). But if you export `REQUIRED_REGEXPS` only and no job is matched then it will simply ignore that fact (which is error prone).

At glance, one solution for `REQUIRED_REGEXPS` would be to track whether all regexps match at least one job or not.

For what I've seen from gh API, job statuses is "queued", "in_progress", "completed", "waiting", "requested" and "pending". This script will threat any not "completed" as RUNNING. So that if there is a required job it will appear on the first query of `get_jobs_for_workflow_run()`   
 
@stevenhorsman do you want this tackled on the first merged version of this solution? I'm okay to leave it as a future improvement.

wainersm reviewed (approved):
@ldoktor thank you so much for this! I hope it's just the begin to make CI experience better.

We still have to include more rules on gatekeeper as well as make the improvements (and fixes) discussed on this PR. However, IMO it's good enough for a kick off. If we don't mark the gatekeeper job required and current rules being loose, then it won't mess with current CI way.

stevenhorsman commented:
> For what I've seen from gh API, job statuses is "queued", "in_progress", "completed", "waiting", "requested" and "pending". This script will threat any not "completed" as RUNNING. So that if there is a required job it will appear on the first query of `get_jobs_for_workflow_run()`
> 
> @stevenhorsman do you want this tackled on the first merged version of this solution? I'm okay to leave it as a future improvement.

Yeah, that's fine as a future improvement

stevenhorsman commented:
So I tried
```
REQUIRED_JOBS="Commit Message Check" REQUIRED_REGEXPS=".*" GITHUB_REPOSITORY=ildikov/kata-containers COMMIT_HASH=5e03bec26b5d40b48c60ceaa5eda5f533ea34842 python3 tools/testing/gatekeeper/jobs.py
```
with the new code and it seemed to just get stuck:
```
WARN: Commit Message Check - Still running

Some jobs are still running.
1 jobs are still running...
WARN: Commit Message Check - Still running

Some jobs are still running.
1 jobs are still running...
WARN: Commit Message Check - Still running
...
```
but that branch is pretty old now and so the fact it doesn't syntax error is probably enough

stevenhorsman reviewed (approved):
I think there are still a few improvements to go on this, but it's a great start, so I'm happy to see it merged and start getting used in action. Thanks @ldoktor

gkurz reviewed (commented):
Can we see some test runs of these changes before merging ?
i.e. push this PR to a dummy branch in the kata-containers org and _replay_ some selected PRs on top ?

gkurz reviewed (dismissed):
> Can we see some test runs of these changes before merging ? i.e. push this PR to a dummy branch in the kata-containers org and _replay_ some selected PRs on top ?

Lukas is away this week. I'll run the smoke tests myself.

Blocking merge in the meantime.

wainersm commented:
Hi @gkurz !

> Can we see some test runs of these changes before merging ? i.e. push this PR to a dummy branch in the kata-containers org and _replay_ some selected PRs on top ?

I'm afraid it's not as simple as that to get it tested. IIRC @ldoktor used his own org/repo and made some mini-CI for testing it.

I linted the workflows that I fixed the conflicts.

If we don't mark the gatekeeper job required then I don't think it will block anyone PR in case gatekeeper fails.

Yes, I'm suggesting to test it on production. If something goes terrible wrong we can git-revert it :)

gkurz commented:
> Hi @gkurz !
> 
> > Can we see some test runs of these changes before merging ? i.e. push this PR to a dummy branch in the kata-containers org and _replay_ some selected PRs on top ?
> 
> I'm afraid it's not as simple as that to get it tested. IIRC @ldoktor used his own org/repo and made some mini-CI for testing it.
> 
> I linted the workflows that I fixed the conflicts.
> 
> If we don't mark the gatekeeper job required then I don't think it will block anyone PR in case gatekeeper fails.
> 
> Yes, I'm suggesting to test it on production. If something goes terrible wrong we can git-revert it :)

I created https://github.com/kata-containers/kata-containers/tree/selective-ci as a clone of this PR.
I then created PR https://github.com/kata-containers/kata-containers/pull/9961 against this branch and added `ok-to-test`.
I noticed that gatekeeper was stuck on waiting for the commit message check to pass.
The commit message check was failing indeed. I've fixed it and forced-push but the gatekeeper remained stuck.
I cancelled it and restarted it. Same outcome. Stuck at 
https://github.com/kata-containers/kata-containers/actions/runs/9784217435/job/27015744800

wainersm commented:
Hi @gkurz ! First of all, thanks for testing it!

I sent a small fix for forcing the running job to cancel on case of updates on the PR

I'm going to still look at the problem of gatekeeper getting stuck. Stay tuned!

gkurz commented:
> Hi @gkurz ! First of all, thanks for testing it!
> 
> I sent a small fix for forcing the running job to cancel on case of updates on the PR
> 

Thanks a lot @wainersm ! I gave a try and the job now gets cancelled as expected.

> I'm going to still look at the problem of gatekeeper getting stuck. Stay tuned!

And it is still ending up stuck, all the time. Worth to note that `Static checks self-hosted / skipper / skipper` and `Static checks self-hosted / build-checks` checks get cancelled right away when CI gets restarted.

wainersm commented:
Hi @gkurz !

I sent a workaround for the problem of gatekeeper being stuck. I changed the name of "Commit Message Check"  to "Commit Message Check / Commit Message Check", because in https://github.com/kata-containers/kata-containers/pull/9637/commits/b546a0413d66b7f82af29084f3a8237432f640db#diff-1d2a599a33f4c06035386878f918fe1d795bb847982c256b67dc4ea8ecd8fadbR66 it compares with workflow_name + job_name. In other words, the list of `required_tests` in `required-tests.yaml` must be of the form `workflow name / job name`.

I tried to fix the problem discussed in https://github.com/kata-containers/kata-containers/pull/9637#issuecomment-2204288031 but I failed miserably....it seems not an easy to solve problem as the gatekeeper does run asynchronously with the tests.

wainersm commented:
Sent another fix. COMMIT_HASH should be exported in the environment.

ldoktor commented:
Changes of https://github.com/kata-containers/kata-containers/compare/748ec286aad0f550010f3eebc357d663f053a718..d8cde12b252ec010542cd14c979edcfdf60c6b2e

* Fixed long line in `workflows/gatekeeper: export COMMIT_HASH variable`
* New commit to unify separators for multiple regexps passing
* New commit to check required labels before checking the test results
* New commit to add extra debug output

I deliberately not rebased it to allow simpler diff review, I'll rebase it now with no changes...

ldoktor commented:
Changes:

* Just rebased, no other changes from [d8cde12](https://github.com/kata-containers/kata-containers/pull/9637/commits/d8cde12b252ec010542cd14c979edcfdf60c6b2e)

ldoktor commented:
Changes:

* Fixed typo (extra `$` character in output) in https://github.com/kata-containers/kata-containers/pull/9637/commits/8cfa92128e5764f2aa283e7a5c13cdd195685767
* New commit that ensures gatekeeper uses the latest workflow results (previously it used the last one to be retrieved which was random)

ldoktor commented:
Changes:

* Fixed an incorrectly rebased `ci: Reshuffle run-k8s-tests to avoid "too many workflows"` which contained two `if` conditionals for one workflow
* New commit to replace `workflow` for `workflow_sha` to avoid dependency deadlock

ldoktor commented:
Changes:

* Replaced the workflow_sha with workflow_ref as workflow_sha is shared across all workflows (last commit amended)

ldoktor commented:
Changes:

* Removed the last commit as the `workflow` concurrency name was not the cause of the problems
* Amended `workflows/gatekeeper: configure concurrency properly` commit to not to set concurrency for `gatekeeper-skipper` workflow as it's a called job with set group from the caller (which was causing the deadlock in concurrency selector)

ldoktor commented:
Well, the testing results are here: https://github.com/kata-containers/kata-containers/pull/9979 and the status seems accurate. The `make static-checks` faild because of a broken link as well as a couple of other tests, the gatekeeper detected this and reported all passed tests, several tests that are still running and all the currently failed tests. Then it interrupted as there is no point in keep monitoring the status and reported a failure.

Greg left on PTO so @wainersm @stevenhorsman would you please take a look at the current version?

wainersm commented:
Hi @ldoktor !

> Well, the testing results are here: #9979 and the status seems accurate. The `make static-checks` faild because of a broken link as well as a couple of other tests, the gatekeeper detected this and reported all passed tests, several tests that are still running and all the currently failed tests. Then it interrupted as there is no point in keep monitoring the status and reported a failure.
> 
> Greg left on PTO so @wainersm @stevenhorsman would you please take a look at the current version?

I stand my Ack.

However, this will conflict with two that are about to get merged: https://github.com/kata-containers/kata-containers/pull/10007 and https://github.com/kata-containers/kata-containers/pull/10038 . We just need to sync up the order things will get in. I think 10007 and 10038 are priority. @sprt wdyt?

Also on 10007 I did some simplifications on the workflows, maybe I can cherry-pick-and-modify https://github.com/kata-containers/kata-containers/pull/9637/commits/62ffd83573795199012ac13d34d4868e6150a6f2

ldoktor commented:
Thanks, please let me know when I should rebase.

wainersm commented:
Hi @ldoktor ! https://github.com/kata-containers/kata-containers/pull/10007 was merged, so you can rebase this PR now.

ldoktor commented:
Changes:

* Rebased
* Removed 62ffd83573795199012ac13d34d4868e6150a6f2 `ci: Reshuffle run-k8s-tests to avoid "too many workflows"` as hopefully Wainer's changes decreased the amount of checks enough (need to run the ci to see)

BbolroC commented on .github/workflows/gatekeeper-skipper.yaml:
```diff
@@ -0,0 +1,52 @@
+          TARGET_BRANCH: ${{ inputs.target-branch }}
+        run: |
+          python3 tools/testing/gatekeeper/skips.py | tee -a "$GITHUB_OUTPUT"
+        shell: /usr/bin/bash -x {0}
```
I was wondering if the placeholder `{0}` is necessary here. I believe the `run:` block automatically executes the provided command in the specified shell. Just curious. Thanks!

ldoktor replied:
TBH I don't know. I copied it from somewhere while working on something else and never experiment with it since it worked well.

BbolroC replied:
yeah, then let's keep it the way it is. Thanks!

BbolroC commented on tools/testing/gatekeeper/skips.py:
```diff
@@ -0,0 +1,109 @@
+    def get_features(self, target_branch):
+        """
+        Get changed file to `target_branch` and map them to the
+        to-betested set-of-tests
```
I would expect `to-be-tested` here, but if I'm mistaken, feel free to ignore this. Thanks!

ldoktor replied:
ack

BbolroC commented on tools/testing/gatekeeper/skips.py:
```diff
@@ -0,0 +1,109 @@
+        print('\n'.join(changes), file=sys.stderr)
+        enabled_features = set()
+        # Go through lines and find what features should be covered
+        for line in changes:
```
How about using `changed_files` and `changed_file` instead of `changes` and `line` to improve readability? Thanks!

ldoktor replied:
ack

BbolroC commented on tools/testing/gatekeeper/required-tests.yaml:
I was wondering if the changed behaviour of the CI (based on this configuration) could at least meet our CI requirement after this PR with this configs and rules is merged. Or are you waiting for a full legit set of configs and rules from the community? Thanks!

I was wondering if the changed behavior of the CI based on this configuration would still meet our CI requirements after this PR is merged. Or are we waiting for a full, finalized set of configs and rules from the community? Thanks!

ldoktor replied:
it definitely does not covers the current list of required tests but it should serve as an example. Before removing any required tests this should be proven to work well and a commit to cover the required tests should be merged. I can help with it but mostly I'd expect the community to send PRs to the expected tests before removing them from the list of requires.

(similarly maintainers of subsystems should contribute the "ignores" to avoid executing tests that do not make sense)

BbolroC replied:
Thanks for the explanation. I’ve confirmed that the current `required-tests.yaml`’s `paths` section accounts for the following result for all changed files:

```
skip_static=no
skip_build=no
skip_test=no
```

However, for files under ci/openshift-ci and .github/workflows/, the script returns:

``
`
skip_static=yes
skip_build=yes
skip_test=yes
```

The current configuration for the mapping section only dictates how to set a required job and doesn’t alter the CI workflow itself. I agree with your perspective, and future PRs for the config should be raised in a way that doesn’t impact the CI. Thanks again.

BbolroC commented on tools/testing/gatekeeper/required-tests.yaml:
```diff
@@ -0,0 +1,42 @@
+  test:
+    # Checks the basic functional tests work
+    regexps: "Kata Containers CI / .*run-basic-amd64-tests.*|Kata Containers CI / .*run-metrics-tests.*"
+    names:
```
How about using `tests` here to match `required_tests` (similar to the pair `required_regexps` and `regexps`)

ldoktor replied:
I could do but given the file is called `required-tests` I assumed this prefix by default (but maybe I'm just too lazy to type :-) )

ldoktor replied:
Just to explicitly state, I'm not changing this for the next version unless you (or someone) asks me to do so.

BbolroC commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,250 @@
+                self.required_regexps.append(re.compile(regexp))
+        if not required_jobs and not self.required_regexps:
+            raise RuntimeError("No REQUIRED_JOBS or REQUIRED_REGEXPS defined")
+        # Set all required jobs as RUNNING to enforce waiting for them
```
I'm seeing the status `EXPECTED` below. The comment here is a bit confusing. Am I missing something? Thanks!

ldoktor replied:
sure, let me change the comment...

BbolroC commented on tools/testing/gatekeeper/jobs.py:
```diff
@@ -0,0 +1,250 @@
+        for job in self.results.values():
+            status = self._job_status(job)
+            if status == RUNNING:
+                running |= True
```
Is there any particular reason for using `running |= True` instead of the more intuitive `running = True`?

ldoktor replied:
Not really, I'm just used to this from before as it (used to be) faster.

BbolroC reviewed (commented):
Thanks for the great work! I just went over the PR, and it's amazing—I learned a lot. I’ve left a few quick questions if you’d like to check them out. Thanks, @ldoktor! ❤️

ldoktor commented:
Changes:

* no changes, just rebase (the conflict was in ubuntu-latest -> ubuntu-22.04

**Do not review yet, I'll incorporate @BbolroC's comments shortly but I wanted to separate the commits to simplify reviewing the changes**

ldoktor commented:
Changes:

* A few variable renames, typo fixes and comment refinements

Not changed:

* Yaml file definition (feel free to prove me wrong but given the file name I'd go with the shorter version)

stevenhorsman commented:
I think this is ready to merge now. @ldoktor do you want to do the honours?

ldoktor commented:
> I think this is ready to merge now. @ldoktor do you want to do the honours?

Thanks but I can't, no permissions to do so...

gkurz commented:
> > I think this is ready to merge now. @ldoktor do you want to do the honours?
> 
> Thanks but I can't, no permissions to do so... (and besides, the reviewer is responsible for the code quality/functionality so merging it myself would put me into a bad position ;-) )

Right :wink:
